### PR TITLE
Forward html content type only

### PR DESCRIPTION
**Pull request by tuzz**
See [here](https://github.com/alphagov/publishing-api/blob/b532ef8c883598cc5faa91e45d4bba53a173bca2/doc/representation-for-multiple-content-types.md) for a decision record regarding the representation.

**Post deploy:**

- [ ] Document multiple representations in the content schemas
- [ ] Document multiple representations in the syntax / semantics docs
- [ ] Create a pull request to remove old_description columns

https://trello.com/c/nkofgTdj/372-add-raw-govspeak-to-publishing-api-payload

### CODE REVIEW

File: app/models/draft_content_item.rb [hunk 1]
```diff
@@ -35,6 +35,29 @@ class DraftContentItem < ActiveRecord::Base
...
+
+  def description
+    value = super
+
+    if value.is_a?(Hash) && value.key?(:string)
+      value.fetch(:string)
+    else
+      value
+    end
+  end
```
Comment (tuzz) [inline app/models/draft_content_item.rb]: If anyone has ideas on how to make this better – I'd appreciate it.

I had a quick poke around [in Rails](https://github.com/rails/rails/blob/7f18ea14c893cb5c9f04d4fda9661126758332b5/activerecord/lib/active_record/type/internal/abstract_json.rb#L13), but stopped as it began to spiral into a rabbit hole.

I had a quick go at upgrading Rails to 4.2.5 (we're currently 4.2.4), but this didn't help. We should probably do this upgrade anyway. I may put in a pull request.

File: app/presenters/content_type_resolver.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        array.map { |e| resolve(e) }
+      end
+    end
+
+    def has_content_types?(array)
+      return false unless array.all? { |v| v.is_a?(Hash) }
+
+      hash_keys = array.flat_map(&:keys).map(&:to_sym)
+      hash_keys.include?(:content_type)
+    end
```
Comment (tuzz) [inline app/presenters/content_type_resolver.rb]: This method is duplicated in the validator. Ideas for DRYing this up would be appreciated.

File: app/validators/well_formed_content_types_validator.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      array.all? { |e| validate!(e) }
+    end
+  end
+
+  def has_content_types?(array)
+    return false unless array.all? { |v| v.is_a?(Hash) }
+
+    hash_keys = array.flat_map(&:keys).map(&:to_sym)
+    hash_keys.include?(:content_type)
+  end
```
Comment (tuzz) [inline app/validators/well_formed_content_types_validator.rb]: This method is duplicated in the resolver. Ideas for DRYing this up would be appreciated.

File: db/schema.rb [hunk 4]
```diff
@@ -59,7 +60,7 @@
     t.string   "locale",                default: "en"
     t.string   "base_path"
     t.string   "title"
-    t.string   "description"
+    t.string   "old_description"
```
Comment (tuzz) [inline db/schema.rb]: I'm keeping the `old_description` fields for now (just in case). We can remove in another pull request.

File: spec/requests/content_item_requests/message_bus_spec.rb [hunk 5]
```diff
@@ -141,4 +141,73 @@
...
+          transmitted_at: DateTime.now.to_s(:nanoseconds),
+        ).to_json
+
+        delivery_info, _, payload = wait_for_message_on(@queue)
+
+        expect(delivery_info.routing_key).to eq("guide.major")
+        expect(JSON.parse(payload)).to eq(JSON.parse(expected_payload))
+      end
+    end
+  end
```
Comment (tuzz) [inline spec/requests/content_item_requests/message_bus_spec.rb]: This test coverage was missing previously.

File: app/models/draft_content_item.rb (hunk 1 above)
Comment (h-lame) [inline app/models/draft_content_item.rb]: I might be inclined to remove the type check:

```ruby
def description=(value)
  super(string: value.to_s)
end

def description
  super.fetch(:string)
end
```

You lose some correctness for cases where the description is being set as a properly constructed hash already, or it was somehow stored as an incorrect hash.  But, the api of the object is that the description attribute is a string and the content item will store that appropriately - if you give it anything else it'll get massaged into a shape we expect (e.g. what AR would do with any other field that you give it the wrong type of object - it'll do a simple coercion and serialize away).

File: doc/representation-for-multiple-content-types.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
+## Decision Record: Representation for Multiple Content Types
```
Comment (rboulton) [inline doc/representation-for-multiple-content-types.md]: This file looks great.  Would it be better in doc/arch/ to live with the existing ADR that's there?

File: app/models/draft_content_item.rb (hunk 1 above)
Comment (tuzz) [inline app/models/draft_content_item.rb]: Good plan: https://github.com/alphagov/publishing-api/commit/dafa1de03d8d4b4eab63b9eaa9c886f2816a078f

That field can hold other things, that aren't just strings, so I called the key `value`.

I also overrode `#attributes`. See the commit message for more info.

What do you think?

File: doc/representation-for-multiple-content-types.md (hunk 6 above)
Comment (tuzz) [inline doc/representation-for-multiple-content-types.md]: Good plan. Addressed.

File: app/models/draft_content_item.rb (hunk 1 above)
Comment (danielroseman) [inline app/models/draft_content_item.rb]: My only concern with this is that it could make it harder to index this data if we ever need to search it. This may not be a problem in practice though.

File: db/migrate/20151119100750_change_description_column_types.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+class ChangeDescriptionColumnTypes < ActiveRecord::Migration
+  def change
+    [
+      [:draft_content_items, DraftContentItem],
+      [:live_content_items, LiveContentItem],
+    ].each do |table, model|
+      add_column table, :new_description, :json, default: { value: nil }
```
Comment (danielroseman) [inline db/migrate/20151119100750_change_description_column_types.rb]: Given the discussions we've had about migrations which lock tables, and the advice here: https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/ we should probably split this into two actions.

File: doc/arch/representation-for-multiple-content-types.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+## Decision
+
+The decision was made to choose Option 1. This representation is not very
+compact, but it means that JSON objects could store additional information if
+required, such as a Govspeak version. It was noted that this option may not
+be the easiest for Publishing Applications to work with.
+
+Option 2 was refuted on the basis that it doesn't clearly demonstrate which
```
Comment (danielroseman) [inline doc/arch/representation-for-multiple-content-types.md]: quibble: refuted is the wrong word here (it means "disproved"); maybe "rejected"?

File: app/presenters/content_type_resolver.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
+module Presenters
+  class ContentTypeResolver
+    def initialize(content_type)
+      self.content_type = content_type
```
Comment (danielroseman) [inline app/presenters/content_type_resolver.rb]: Why `self.` rather than `@`?

File: spec/requests/content_item_requests/message_bus_spec.rb [hunk 10]
```diff
@@ -141,4 +141,73 @@
...
+
+  describe "content type inlining" do
```
Comment (danielroseman) [inline spec/requests/content_item_requests/message_bus_spec.rb]: Are we certain that this is what we want, ie only the HTML version goes on the queue?

**Comment by danielroseman** [comment]
A few mostly minor comments, plus needs rebasing.

File: app/presenters/content_type_resolver.rb (hunk 9 above)
Comment (tuzz) [inline app/presenters/content_type_resolver.rb]: Two reasons:

1) I'm `nil` averse. If you misspell an instance variable, it's set to `nil` at runtime, which can lead to some nasty / hard-to-find bugs.

2) The object better encapsulates those fields. Contrived example:

Say we wanted to restrict it such that `content_type` must be a valid MIME type from a whitelist. By using a setter method on the object, we can put that logic in a single place rather than have to check it whenever we set the `content_type` on the object.

I try not to use instance variables anywhere if I can help it (except perhaps memoization).

File: spec/requests/content_item_requests/message_bus_spec.rb (hunk 10 above)
Comment (tuzz) [inline spec/requests/content_item_requests/message_bus_spec.rb]: Not certain. Currently, Rummager strips tags from HTML prior to indexing. I spoke to @elliotcm about this and we thought that it's probably ok for it to only send the HTML content type forward, but happy to discuss further.

It could cause problems if we're "throwing information away" and then later find that we need it for some reason.

Comment (rboulton) [inline spec/requests/content_item_requests/message_bus_spec.rb]: Quick correction - rummager doesn't do any HTML tag stripping - but I think some publisher apps which are sending content to rummager do before they send it.

Regardless, I believe having only the HTML is fine for indexing - but we definitely need to know if a field is expected to contain HTML markup, so we can strip it out (or possibly make use of it to index some bits of text as being more important - eg, headings).

I think the current design is that we'd be able to find that out by reading the schema, but not simply by inspecting the document itself without reference to the schema. This is probably fine, though does make it a bit hard to manage if the schema changes from a field being a plain text field to being a govspeak + html field - there's no way for our indexing code to know about the change, and thus to ensure that it's stripping out HTML tags for all HTML fields.  This is probably something we'll just have to live with, though.

File: spec/presenters/content_type_resolver_spec.rb [hunk 11]
```diff
@@ -0,0 +1,97 @@
+require "rails_helper"
+
+RSpec.describe Presenters::ContentTypeResolver do
+  subject { described_class.new("html") }
+
+  it "inlines content of the specified content type" do
```
Comment (rboulton) [inline spec/presenters/content_type_resolver_spec.rb]: Could you add a test of what happens if the `html` version isn't present?  ie, if it gets
`body: [ { content_type: "text", content: "some content" } ]`

I'm not sure if it should return `"body": null`, or omit `body` entirely.  It definitely shouldn't fall back to any other representation (because then downstream apps won't know what content type the data they get is).

Comment (tuzz) [inline spec/presenters/content_type_resolver_spec.rb]: I designed it such that the system fails early on receiving content that isn't "well formed" through a custom validator. The scenario you describe is tested here:

https://github.com/alphagov/publishing-api/blob/forward-html-content-type-only/spec/validators/well_formed_content_types_validator_spec.rb#L67-L76

My thinking was that I'd fail early on receiving illy formed content and relieve the ContentTypeResolver of this responsibility. I'm happy to double test this though – but I'd be a little concerned that people would think it's acceptable to send illy formed content to the resolver.

What do you think?

File: spec/requests/content_item_requests/message_bus_spec.rb (hunk 10 above)
Comment (elliotcm) [inline spec/requests/content_item_requests/message_bus_spec.rb]: @rboulton Is there any harm to just stripping HTML on all fields?

**Comment by elliotcm** [comment]
@tuzz Looks fine, just need @rboulton's sign-off on a couple of things and needs the migration splitting as per @danielroseman's comment.

**Comment by tuzz** [comment]
All comments addressed, waiting on feedback from @rboulton on his points and confirmation from @danielroseman that the migration split is acceptable.

On this point, I've tested it against preview data with:

```ruby
DraftContentItem.find_each.map(&:description)
#=> [ ... ]

LiveContentItem.find_each.map(&:description)
#=> [ ... ]
```

Working as expected.

**Comment by danielroseman** [comment]
:+1:, needs rebasing though.

**Comment by tuzz** [comment]
Rebased.

File: spec/requests/content_item_requests/message_bus_spec.rb (hunk 10 above)
Comment (rboulton) [inline spec/requests/content_item_requests/message_bus_spec.rb]: Stripping all HTML would be a problem, because we are quite likely to want to be able to use things like <h1> tags to mark some parts of the text as more important.  Plus, some fields may contain identifiers with characters like < which are significant.

File: spec/presenters/content_type_resolver_spec.rb (hunk 11 above)
Comment (tuzz) [inline spec/presenters/content_type_resolver_spec.rb]: Any thoughts @rboulton ?

File: spec/requests/content_item_requests/message_bus_spec.rb (hunk 10 above)
Comment (tuzz) [inline spec/requests/content_item_requests/message_bus_spec.rb]: I see. We will have to bear that in mind when we tackle the consumer of those messages on the queue. Thanks for clearing that up.

Given this point is somewhat orthogonal to the code in this pull request – do you have any objections if we merge this? It's becoming stale and we're keen to get this through.

**Comment by tuzz** [comment]
Following a discussion with @danielroseman, @bradwright and @elliotcm – we're going to close this and take a slightly different approach. We will send all representations downstream and resolve the 'html' content type in a presenter in the content store.
